### PR TITLE
feature: deferred loading and requirement pruning

### PR DESCRIPTION
**Pull request by leondz**
Support construction-time loading of optional modules. Includes

* many generators now using this pattern
* pattern for loading modules at run-time and failing if absent
* optional requirements moved to pyproject.toml options **and pruned from requirements.txt**
* `_load_deps` and `_clear_deps` pattern, used in generator constructor and `_load_client` / `_clear_client`
* tests to check that optional deps are in the right place and not in the wrong place

### CODE REVIEW

File: garak/generators/nemo.py [hunk 1]
```diff
@@ -11,12 +11,14 @@
...
-import nemollm
 
 from garak import _config
-from garak.exception import APIKeyMissingError
+from garak.exception import APIKeyMissingError, GeneratorBackoffExceptionPlaceholder
 from garak.generators.base import Generator
 
+nemollm_serversideerror = GeneratorBackoffExceptionPlaceholder
+nemollm_toomanyrequestserror = GeneratorBackoffExceptionPlaceholder
+
```
[the remark this answers is not in the document]
Comment (leondz) [reply garak/generators/nemo.py:21]: thanks. this defers resolution til runtime, i like it

File: pyproject.toml [hunk 2]
```diff
@@ -115,6 +107,39 @@ lint = [
...
+]
+plugin_ollama = [
+  "ollama>=0.4.7",
+]
+plugin_optimum = [
+  "optimum-nvidia",
+]
+plugin_replicate = [
+  "replicate>=0.8.3",
+]
```
Comment (jmartin-tech) [inline pyproject.toml:142]: Consider adding to allows for `pip install garak[plugin_all]`:
```
plugin_all = garak[
    plugin_cohere,
    plugin_langchain,
    plugin_llava,
    plugin_litellm,
    plugin_mistralai,
    plugin_nemoguardrails,
    plugin_nemollm,
    plugin_octoai,
    plugin_ollama,
    plugin_replicate,
]
```

There might be a better name to use for this group.

File: requirements.txt
Comment (jmartin-tech) [inline requirements.txt:1]: Should we keep `requirements.txt` as a consolidated lists of all possible requirements and require usage of `pyproject.toml` to get a limited dependencies install?

Another option to me would be simply remove `requirements.txt` and require all installs be based on using `pyproject.toml` until we implement a lock file.

Comment (leondz) [reply requirements.txt:1]: Right. After migrating the gh workflows to use pyproject.toml, requirements.txt becomes superfluous - I'm leaning toward the latter. 

Removal does mean we lose a cross check in test_reqs, but that's OK. Could do with a test to ensure all packages (including optional) specity a version.

File: garak/_plugins.py [hunk 3]
```diff
@@ -402,6 +402,23 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+            "extra_dependency_names"
+        ]
+        if len(extra_dependency_names) > 0:
+            for dependency_module_name in extra_dependency_names:
+                for dependency_path in [ # support both plain names and also multi-point names e.g. langchain.llms
+                    ".".join(dependency_module_name.split(".")[: n + 1])
+                    for n in range(dependency_module_name.count(".") + 1)
+                ]:
+                    if importlib.util.find_spec(dependency_path) is None:
+                        _import_failed(dependency_path, full_plugin_name)
```
[the remark this answers is not in the document]
Comment (erickgalinkin) [reply garak/_plugins.py:420]: Oh, I guess that is what we're doing. This is the hazard of doing code reviews linearly, I suppose.

File: garak/generators/huggingface.py [hunk 4]
```diff
@@ -158,19 +159,15 @@ class OptimumPipeline(Pipeline, HFCompatible):
     generator_family_name = "NVIDIA Optimum Hugging Face 🤗 pipeline"
     supports_multiple_generations = True
     doc_uri = "https://huggingface.co/blog/optimum-nvidia"
+    extra_dependency_names = ["optimum-nvidia"]
```
Comment (erickgalinkin) [inline garak/generators/huggingface.py:162]: Minor note that has little to do with this PR: it does drive me a bit nuts that the dependency name and the import statement so often do not match.

File: garak/generators/huggingface.py [hunk 5]
```diff
@@ -70,6 +69,7 @@ def __init__(self, name="", config_root=_config):
         self._load_client()
 
     def _load_client(self):
+        self._load_deps()
         if hasattr(self, "generator") and self.generator is not None:
             return
```
Comment (erickgalinkin) [inline garak/generators/huggingface.py:74]: In the interest of DRYness, I notice this exact code repeated across a number of the `HFCompatible` classes; perhaps we should refactor into `HFCompatible`?

File: garak/generators/langchain.py [hunk 6]
```diff
@@ -53,14 +50,7 @@ def __init__(self, name="", config_root=_config):
 
         super().__init__(self.name, config_root=config_root)
```
Comment (erickgalinkin) [inline garak/generators/langchain.py:52]: Are we missing a call to `self._load_deps()`?

File: garak/generators/litellm.py [hunk 7]
```diff
@@ -105,6 +96,7 @@ class LiteLLMGenerator(Generator):
         "skip_seq_start",
         "skip_seq_end",
         "stop",
+        "verbose",
     )
 
     def __init__(self, name: str = "", generations: int = 10, config_root=_config):
```
Comment (erickgalinkin) [inline garak/generators/litellm.py:102]: Are we missing a call to `self._load_deps()`?

File: garak/generators/nemo.py [hunk 8]
```diff
@@ -37,6 +36,7 @@ class NeMoGenerator(Generator):
 
     supports_multiple_generations = False
     generator_family_name = "NeMo"
+    extra_dependency_names = ["nemollm"]
 
     def __init__(self, name=None, config_root=_config):
```
Comment (erickgalinkin) [inline garak/generators/nemo.py:41]: Are we missing a call to `self._load_deps()`? 

I keep asking this but I may be missing some logic that's inherited, so my question is very sincere.

File: garak/generators/octo.py [hunk 9]
```diff
@@ -30,38 +30,46 @@ class OctoGenerator(Generator):
...
+    extra_dependency_names = ["octoai-sdk"]
 
     def __init__(self, name="", config_root=_config):
-        from octoai.client import Client
 
         self.name = name
         self._load_config(config_root)
         self.fullname = f"{self.generator_family_name} {self.name}"
 
         super().__init__(self.name, config_root=config_root)
```
Comment (erickgalinkin) [inline garak/generators/octo.py:41]: Are we missing a call to `self._load_deps()`?

File: garak/generators/ollama.py [hunk 10]
```diff
@@ -28,17 +32,18 @@ class OllamaGenerator(Generator):
     active = True
     generator_family_name = "Ollama"
     parallel_capable = False
+    extra_dependency_names = ["ollama"]
 
     def __init__(self, name="", config_root=_config):
```
Comment (erickgalinkin) [inline garak/generators/ollama.py:37]: Are we missing a call to `self._load_deps()`?

I'm going to stop asking every time and assume I've missed some inherited logic.

File: garak/generators/openai.py [hunk 11]
```diff
@@ -158,6 +158,7 @@ def __setstate__(self, d) -> object:
     def _load_client(self):
         # When extending `OpenAICompatible` this method is a likely location for target application specific
         # customization and must populate self.generator with an openai api compliant object
+        self._load_deps()
```
Comment (erickgalinkin) [inline garak/generators/openai.py:161]: We don't seem to have added the `openai` dependency as `extra_dependency_names`. Is that deliberate?

**Review by erickgalinkin — COMMENTED** [review]
Looks good to me overall! Just a handful of comments throughout.

File: garak/generators/openai.py (hunk 11 above)
Comment (jmartin-tech) [reply garak/generators/openai.py:163]: OpenAI is among the most commonly used generators and has been kept as a required _out of the box_ enabled generator.

File: garak/generators/base.py [hunk 12]
```diff
@@ -63,6 +65,29 @@ def __init__(self, name="", config_root=_config):
...
+                    extra_dep_name,
+                    garak._plugins.load_optional_module(extra_dependency),
+                )
+
+    def _clear_deps(self):
+        # unload external dependencies from class. should be invoked before
+        # serialisation, esp. in _clear_client (if used)
+        for extra_dependency in self.extra_dependency_names:
+            extra_dep_name = extra_dependency.replace(".", "_")
+            setattr(self, extra_dep_name, None)
```
[the remark this answers is not in the document]
Comment (leondz) [reply garak/generators/base.py]: definitely makes sense to factor it up, thanks

File: garak/generators/huggingface.py (hunk 4 above)
Comment (leondz) [reply garak/generators/huggingface.py]: yes. pillow/PIL and octo/octo-ai/octo-sdk are also culprits here.

File: .github/workflows/maintain_cache.yml [hunk 13]
```diff
@@ -43,7 +43,7 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          pip install -r requirements.txt
+          pip install .
```
Comment (jmartin-tech) [inline .github/workflows/maintain_cache.yml]: This should install all dependencies as the cache file needs to include all plugins.

File: .github/workflows/remote_package_install.yml
Comment (jmartin-tech) [inline .github/workflows/remote_package_install.yml:1]: Consider adding another test here for `all` dependencies install or possibly adding a set of tests that validates each `plugin` group of dependencies.

File: pyproject.toml (hunk 2 above)
Comment (jmartin-tech) [reply pyproject.toml:149]: What happens when we have multiple groups with the same dependency?  I am concerned this could end up creating a maintenance issue if two groups provided a diverging requirements set.

File: tests/buffs/test_buffs.py [hunk 14]
```diff
@@ -21,11 +21,14 @@ def test_buff_structure(classname):
...
     if c._supported_params is not None:
-        if hasattr(g, "DEFAULT_PARAMS"):
+        if hasattr(c, "DEFAULT_PARAMS"):
             for k, _ in c.DEFAULT_PARAMS.items():
                 if k not in c._supported_params:
                     unsupported_defaults.append(k)
     assert unsupported_defaults == []
+    assert hasattr(c, "extra_dependency_names") and isinstance(
+        c.extra_dependency_names, list
+    ), "extra_dependency_names must be a list"
```
Comment (jmartin-tech) [inline tests/buffs/test_buffs.py]: This is a common requirement for all plugins types, should this move into `test_plugins.py` which iterates all plugins?

File: tests/generators/test_generators.py [hunk 15]
```diff
@@ -112,7 +116,13 @@ def test_instantiate_generators(classname):
...
 
     m = importlib.import_module("garak." + ".".join(classname.split(".")[:-1]))
-    g = getattr(m, classname.split(".")[-1])(config_root=config_root)
+    klass = getattr(m, classname.split(".")[-1])
+    try:
+        g = klass(config_root=config_root)
+    except ModuleNotFoundError:
+        pytest.skip(
+            "dependencies not present; requires " + repr(klass.extra_dependency_names)
+        )
```
Comment (jmartin-tech) [inline tests/generators/test_generators.py:128]: Should the call to `_plugins.enumerate_plugins("generators")` for the initial list of generators just reduce the set reported?

I do like that we get indicator of what is skipped here however I wonder if this belongs in this test.

File: tests/test_reqs.py [hunk 16]
```diff
@@ -36,3 +46,75 @@ def test_requirements_txt_pyproject_toml():
     assert (
         reqtxt_reqs == pyproject_reqs
     )  # final check. this one is actually enough, but let's help us debug by finding which test fails, ok?
+
+
+PLUGIN_TYPES = ["probes", "detectors", "harnesses", "generators", "buffs"]
```
Comment (jmartin-tech) [inline tests/test_reqs.py]:
```suggestion
PLUGIN_TYPES = garak._plugins.PLUGIN_TYPES
```

File: tests/test_reqs.py [hunk 17]
```diff
@@ -36,3 +46,75 @@ def test_requirements_txt_pyproject_toml():
...
+        pyproject_toml = tomllib.load(pyproject_file)
+    optional_deps = set()
+    for test_group in pyproject_toml["project"]["optional-dependencies"]:
+        if test_group.startswith("plugin_"):
+            group_dep_names = [
+                split_out_module_name(r)
+                for r in pyproject_toml["project"]["optional-dependencies"][test_group]
+            ]
+            optional_deps.update(group_dep_names)
+    return optional_deps
```
Comment (jmartin-tech) [inline tests/test_reqs.py]: These seem like expensive functions to call for every test, consider making this a fixture that is scoped to the testing session for this module.

Though maybe these tests just go away as `requirements.txt` may be removed with this PR.

File: tests/generators/test_ollama.py [hunk 18]
```diff
@@ -27,6 +42,15 @@ def no_models():
...
 
+@pytest.mark.skipif(
+    not all(
+        [
+            importlib.util.find_spec(m)
+            for m in OllamaGeneratorChat.extra_dependency_names
+        ]
+    ),
+    reason="missing optional dependency",
+)
```
Comment (jmartin-tech) [inline tests/generators/test_ollama.py:67]: Another place maybe to consolidate as a `fixture` scoped to the test module.

File: garak/exception.py [hunk 19]
```diff
@@ -14,7 +14,7 @@ class ModelNameMissingError(GarakException):
     """A generator requires model_name to be set, but it wasn't"""
 
 
-class GarakBackoffTrigger(GarakException):
+class GeneratorBackoffTrigger(GarakException):
```
Comment (jmartin-tech) [inline garak/exception.py:17]: Why rename this? The original name seems clear enough to me, I there some envisioned case where two layers of backoff would need to differentiate the source?

File: .github/workflows/test_linux.yml [hunk 20]
```diff
@@ -49,7 +49,7 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          pip install --no-cache-dir -r requirements.txt
+          pip install --no-cache-dir .[tests]
```
Comment (jmartin-tech) [inline .github/workflows/test_linux.yml]: Either `tests` needs to install all optional dependencies or we need to have more tasks that run tests for each dependency group.

File: .github/workflows/lint.yml [hunk 21]
```diff
@@ -32,7 +32,7 @@ jobs:
       - name: Install dependencies
         run: |
           python -m pip install --upgrade pip
-          pip install -r requirements.txt
+          pip install .[tests]
```
Comment (jmartin-tech) [inline .github/workflows/lint.yml]: I think linter will want to have all possible dependencies.

File: pyproject.toml (hunk 2 above)
Comment (leondz) [reply pyproject.toml:149]: This seems no worse than status quo, except with optional installation, two plugins with conflicts can now co-exist as descriptions

File: garak/generators/base.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (leondz) [reply garak/generators/base.py]: the generator implementation is special, because generators have client load/unload for probe parallelisation. will slate upfactoring for a second iteration.

File: garak/generators/huggingface.py (hunk 5 above)
Comment (leondz) [reply garak/generators/huggingface.py:75]: `_load_client`, the calling method, is class-specific

File: garak/generators/langchain.py (hunk 6 above)
Comment (leondz) [reply garak/generators/langchain.py:51]: good q. `super()` is `generators.base.Generator` whose constructor includes a call to `_load_deps()`

File: garak/generators/litellm.py (hunk 7 above)
Comment (leondz) [reply garak/generators/litellm.py:104]: last line of this constructor calls parent constructor who calls it

File: garak/generators/octo.py (hunk 9 above)
Comment (leondz) [reply garak/generators/octo.py]: octo plugin deleted

File: garak/generators/nemo.py (hunk 8 above)
Comment (leondz) [reply garak/generators/nemo.py:41]: it's good, thank you. call is in place here

File: garak/generators/ollama.py (hunk 10 above)
Comment (leondz) [reply garak/generators/ollama.py:38]: parent constructor call is in place in class constructor

File: .github/workflows/maintain_cache.yml (hunk 13 above)
Comment (leondz) [reply .github/workflows/maintain_cache.yml]: updated

**Comment by leondz** [comment]
need to solve how optional/deferred testing works

File: .github/workflows/lint.yml (hunk 21 above)
Comment (leondz) [reply .github/workflows/lint.yml]: updated

File: .github/workflows/test_linux.yml (hunk 20 above)
Comment (leondz) [reply .github/workflows/test_linux.yml]: updated

File: garak/exception.py (hunk 19 above)
Comment (leondz) [reply garak/exception.py:17]: two reasons: 
- didn't like the redundant `garak` prefix
- want granularity to scope to generators (agree this is currently speculative, but a generic `BackoffException` seems begging for trouble in a few quarters in e.g. adaptive probes, llmaaj)

File: tests/generators/test_ollama.py (hunk 18 above)
Comment (leondz) [reply tests/generators/test_ollama.py:67]: is the suggested pattern:
* try to do imports in a module level fixture
* have the tests requiring those imports require that fixture

this sounds fine - though does it obviate the need to have test-level `skipif`s predicated on availability of the fixture? or is the pattern something else?

File: tests/test_reqs.py (hunk 17 above)
Comment (leondz) [reply tests/test_reqs.py]: > to call for every test

i think these are just called in these tests?

> Though maybe these tests just go away as requirements.txt may be removed with this PR.

prefer separate PR and then, yes

File: tests/generators/test_generators.py (hunk 15 above)
Comment (leondz) [reply tests/generators/test_generators.py:128]: > just reduce the set reported?

beyond the `active=True` requirement?

there's a tension between wanting good dynamic coverage of generator tests, and having to maintain an allowlist of what's safetly instantiable in tests.

i do want to get /some/ coverage here - at least checking modules for syntax failures or basic instantiability - without the td of an allowlist

File: tests/buffs/test_buffs.py (hunk 14 above)
Comment (leondz) [reply tests/buffs/test_buffs.py]: Will do. Currently thinking this should be applied to `Configurable` classes in core plugins.

There may be other things that belong in a `test_plugins.py` but prefer limiting further scope creep of this PR and doing that later

File: tests/generators/test_ollama.py (hunk 18 above)
Comment (erickgalinkin) [reply tests/generators/test_ollama.py:67]: I'll point you to this example, which I believe is approximately what @jmartin-tech is thinking.
https://github.com/NVIDIA-NeMo/Guardrails/blob/fa113814ffb05ec267f63ae5b862382ac16dcd4c/tests/test_utils.py#L150

**Comment by leondz** [comment]
current test failures seem to be due to cache consistency missing `extra_dependency_names` (they pass when checked out)

File: tests/generators/test_muiltiprocessing_support.py [hunk 22]
```diff
@@ -15,11 +15,11 @@
 GENERATORS = [
     "generators.huggingface.InferenceAPI",
     "generators.huggingface.InferenceEndpoint",
-    "generators.mistral.MistralGenerator",
+#    "generators.mistral.MistralGenerator",
```
Comment (jmartin-tech) [inline tests/generators/test_muiltiprocessing_support.py]: Prefer removal to commented out entries.

File: garak/generators/huggingface.py [hunk 23]
```diff
@@ -168,19 +169,15 @@ class OptimumPipeline(Pipeline, HFCompatible):
...
 
-        try:
-            from optimum.nvidia.pipelines import pipeline
-            from transformers import set_seed
-        except Exception as e:
-            logging.exception(e)
-            raise GarakException(
-                f"Missing required dependencies for {self.__class__.__name__}"
-            )
+        pipeline = self.optimum.nvidia.pipelines.pipeline
```
Comment (jmartin-tech) [inline garak/generators/huggingface.py]: Should this class be providing a `_load_deps()` override?

``` Python
    def _load_deps(self):
        return super()._load_deps(["optimum.nvidia"])
```

and accessing:

```suggestion
        pipeline = self.optimum_nvidia.pipelines.pipeline
```

File: garak/generators/ollama.py [hunk 24]
```diff
@@ -3,16 +3,20 @@
...
 from httpx import TimeoutException
 
 
 def _give_up(error):
-    return isinstance(error, ollama.ResponseError) and error.status_code == 404
+    return (
+        str(error.__class__) == "<class 'ollama._types.ResponseError'>"
+        and hasattr(error, "status_code")
+        and error.status_code == 404
+    )
```
Comment (jmartin-tech) [inline garak/generators/ollama.py:19]: How brittle is this this check?  I suspect it may be better to check the negative case since this class only allows one other backoff type:


```suggestion
    return (
        not isinstance(error.__cause__, TimeoutException)
        and hasattr(error.__cause__, "status_code")
        and error.__cause__.status_code == 404
    )
```

File: garak/_plugins.py [hunk 25]
```diff
@@ -448,3 +471,44 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+                extra_dep_name,
+                load_optional_module(extra_dependency),
+            )
+
+
+def _clear_deps(self):
+    # unload external dependencies from class. should be invoked before
+    # serialisation, esp. in _clear_client (if used)
+    for extra_dependency in self.extra_dependency_names:
+        extra_dep_name = extra_dependency.replace(".", "_")
```
Comment (jmartin-tech) [inline garak/_plugins.py]: Does this also need to account for `-` similar to `_load_deps()`?

**Review by jmartin-tech — CHANGES_REQUESTED** [review]
Some edge case concerns, I suspect some are difficult to test though I think mocking should be viable.

Original test structure comments still apply however are not blocking for moving this forward.

File: tests/generators/test_muiltiprocessing_support.py (hunk 22 above)
Comment (leondz) [reply tests/generators/test_muiltiprocessing_support.py]: Commented all back in & set test to skip entries that can't be processed

File: garak/generators/huggingface.py (hunk 23 above)
Comment (leondz) [reply garak/generators/huggingface.py]: good idea, thanks

File: garak/generators/ollama.py (hunk 24 above)
Comment (leondz) [reply garak/generators/ollama.py:19]: it's sufficiently brittle. accepted, thanks

File: garak/_plugins.py (hunk 25 above)
Comment (leondz) [reply garak/_plugins.py]: it absolutely does

File: garak/generators/huggingface.py (hunk 23 above)
Comment (leondz) [reply garak/generators/huggingface.py]: pruned optimum - it's EOL and currently incompatible due to dependencies

File: garak/generators/replicate.py [hunk 26]
```diff
@@ -56,34 +56,42 @@ def __getstate__(self) -> object:
...
-        self._load_client()
+        self._load_deps()
 
     def _load_client(self):
-        self.client = importlib.import_module("replicate")
+        self._load_deps()
+        self.client = self.replicate
 
     def _clear_client(self):
+        self._clear_deps()
```
Comment (jmartin-tech) [inline garak/generators/replicate.py:66]: Will need to watch these for #1471 related issues.

File: garak/probes/audio.py [hunk 27]
```diff
@@ -61,12 +63,7 @@ def _load_data(self):
                 "Audio Achilles data not found. Downloading from HuggingFace."
             )
 
-            try:
-                import soundfile as sf
-                from datasets import load_dataset
-            except ImportError as e:
-                logging.critical("Missing libraries for audio modules.", exc_info=e)
-                raise GarakException("Missing Libraries for audio modules.")
+            from datasets import load_dataset
```
Comment (jmartin-tech) [inline garak/probes/audio.py:66]: There is a change in behavior here, with `extra_dependency_names` loading `soundfile` it becomes a required dependency for this probe.  Before this PR it was only required if no user provided files already exist in the `data_path`.

Not a blocker, just noting for awareness.

File: garak/_plugins.py [hunk 28]
```diff
@@ -448,3 +471,44 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+        requesting_module = Path(inspect.stack()[1].filename).name.replace(".py", "")
+        _import_failed([module_name], requesting_module)
+    return m
+
+
+def _import_failed(absent_modules: [str], calling_module: str):
+    quoted_module_list = "'" + "', '".join(absent_modules) + "'"
+    module_list = " ".join(absent_modules)
+    msg = f"⛔ Plugin '{calling_module}' requires Python modules which aren't installed/available: {quoted_module_list}"
+    hint = f"💡 Try 'pip install {module_list}' to get missing module."
```
Comment (jmartin-tech) [inline garak/_plugins.py:460]: Just noting, it might be nice to have this reference a `group` in the pyproject.toml.  This may be added in #1475 when the as these groups are introduced.

File: garak/_plugins.py [hunk 29]
```diff
@@ -448,3 +471,44 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+                load_optional_module(extra_dependency),
+            )
+
+
+def _clear_deps(self):
+    # unload external dependencies from class. should be invoked before
+    # serialisation, esp. in _clear_client (if used)
+    for extra_dependency in self.extra_dependency_names:
+        extra_dep_name = extra_dependency.replace(".", "_").replace("-", "_")
+        setattr(self, extra_dep_name, None)
```
Comment (jmartin-tech) [inline garak/_plugins.py]: Location wise it is nice to have these in the `_plugins` package, however is suspect the parameters accepted should not refer to `self`, it might be more clear to reference `inst` or `obj` or `plugin` as these as `self` has a somewhat expected meaning in python signatures.

**Review by jmartin-tech — APPROVED** [review]
Final testing in progress, landing this may present some risk as testing resources for every generator changed is not feasible.

File: garak/_plugins.py (hunk 29 above)
Comment (leondz) [reply garak/_plugins.py]: agree. will rename to e.g. `plugin`, and type as `Configurable`

File: garak/_plugins.py [hunk 30]
```diff
@@ -404,6 +405,28 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+        if len(extra_dependency_names) > 0:
+            absent_modules = []
+            for dependency_module_name in extra_dependency_names:
+                for (
+                    dependency_path
+                ) in [  # support both plain names and also multi-point names e.g. langchain.llms
+                    ".".join(dependency_module_name.split(".")[: n + 1])
+                    for n in range(dependency_module_name.count(".") + 1)
+                ]:
+                    if importlib.util.find_spec(dependency_path) is None:
```
Comment (jmartin-tech) [inline garak/_plugins.py]: Testing shows `find_spec` needs to be provided the runtime package name, currently `dependency_path` entries are the _pypi_ package name.

This can be tested by attempting to load `generators.huggingface.LLaVA`.
```
⛔ Plugin 'generators.huggingface.LLaVA' requires Python modules which aren't installed/available: 'pillow'
💡 Try 'pip install pillow' to get missing module.
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/jemartin/Projects/nvidia/garak/garak/__main__.py", line 14, in <module>
    main()
  File "/home/jemartin/Projects/nvidia/garak/garak/__main__.py", line 9, in main
    cli.main(sys.argv[1:])
  File "/home/jemartin/Projects/nvidia/garak/garak/cli.py", line 596, in main
    generator = _plugins.load_plugin(
                ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jemartin/Projects/nvidia/garak/garak/_plugins.py", line 428, in load_plugin
    _import_failed(absent_modules, full_plugin_name)
  File "/home/jemartin/Projects/nvidia/garak/garak/_plugins.py", line 493, in _import_failed
    raise ModuleNotFoundError(msg)
ModuleNotFoundError: ⛔ Plugin 'generators.huggingface.LLaVA' requires Python modules which aren't installed/available: 'pillow'
```

Comment (jmartin-tech) [reply garak/_plugins.py]: Looking at this code in particular, I don't think this validation is needed here at this time. It may be more appropriate have handle a missing import exception around the module instantiation call.

File: garak/_plugins.py [hunk 31]
```diff
@@ -404,6 +405,28 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+                        absent_modules.append(dependency_module_name)
+            if len(absent_modules):
+                _import_failed(absent_modules, full_plugin_name)
```
Comment (jmartin-tech) [inline garak/_plugins.py]: Since the goal of `load_plugin` is to return an instance no need to check the dependencies early just let the instance creation raise:
```suggestion
```

File: garak/_plugins.py [hunk 32]
```diff
@@ -448,3 +472,49 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+        extra_dep_name = extra_dependency.replace(".", "_").replace("-", "_")
+        if (
+            not hasattr(plugin_instance, extra_dep_name)
+            or getattr(plugin_instance, extra_dep_name) is None
+        ):
+            setattr(
+                plugin_instance,
+                extra_dep_name,
+                load_optional_module(extra_dependency),
+            )
```
Comment (jmartin-tech) [inline garak/_plugins.py]: May want to adjust the `calling_module` value passed but the class name seems reasonable:
```suggestion
            try:
                setattr(
                    plugin_instance,
                    extra_dep_name,
                    importlib.import_module(extra_dependency),
                )
            except ModuleNotFoundError as e:
                _import_failed(plugin_instance.extra_dependency_names, plugin_instance.__class__.__name__)
```

File: garak/_plugins.py [hunk 33]
```diff
@@ -448,3 +472,49 @@ def load_plugin(path, break_on_fail=True, config_root=_config) -> object:
...
+
+
+def load_optional_module(module_name: str):
+    try:
+        m = importlib.import_module(module_name)
+    except ModuleNotFoundError:
+        requesting_module = Path(inspect.stack()[1].filename).name.replace(".py", "")
+        _import_failed([module_name], requesting_module)
+    return m
+
```
Comment (jmartin-tech) [inline garak/_plugins.py]: Not needed:

```suggestion
```

**Review by jmartin-tech — CHANGES_REQUESTED** [review]
Suggestion offered removes the early extra deps check and defers the load and exception to the actual instance creation step.

This pattern is highly coupled to the current `extra_dependency_names` design, however no more so than the original AFAICT. One drawback is this always reports all `extra_dependency_names` in the hint vs only the specific one that triggered the raise.

File: garak/_plugins.py (hunk 30 above)
Comment (leondz) [reply garak/_plugins.py]: > Looking at this code in particular, I don't think this validation is needed here at this time. It may be more appropriate have handle a missing import exception around the module instantiation call.

I believe the intent here is to summarise in one pass a list of all missing module names, which is determined using `find_spec`.

File: garak/_plugins.py (hunk 31 above)
Comment (leondz) [reply garak/_plugins.py]: Is this in tension with the requested feature of listing all missing modules at once rather than piecemeal? I realise the granularity is different, but don't we want to cause the minimum number of user round trips between execution and dep installation?

File: garak/_plugins.py (hunk 33 above)
Comment (leondz) [reply garak/_plugins.py]: Do you have another way of managing this within `_load_deps`?

File: garak/_plugins.py (hunk 31 above)
Comment (jmartin-tech) [reply garak/_plugins.py]: Not really, load_plugin is only called when actually instantiating a plugin, testing in this location we will not evaluate all plugins required for the run.  Since `generators` are the primary plugin type using this pattern just removing this is acceptable for now.

File: garak/_plugins.py (hunk 33 above)
Comment (jmartin-tech) [reply garak/_plugins.py]: Removal note is based on the suggestion code that was accepted in `_load_deps`.  This function is no longer called.

File: garak/_plugins.py (hunk 30 above)
Comment (jmartin-tech) [reply garak/_plugins.py]: This is simply not the right place for this, `load_plugin` is called at instantiation so this is only evaluating for _this_ plugin which is fully processed by `_load_deps` so not needed here.

In the next iteration PR #1475 we might want to add an early preprocessor that takes a comprehensive look at the full run config to determine if all dependencies required for the full run are available however I am thinking that might turn out to be an overly complex goal that may get deferred or shelved in favor of  allowing the run to skip `probes` that happen to be missing dependencies instead for blocking start of the run. More discussion of that can happen in that PR in later.

File: garak/_plugins.py (hunk 33 above)
Comment (leondz) [reply garak/_plugins.py]: cut

File: garak/_plugins.py (hunk 30 above)
Comment (leondz) [reply garak/_plugins.py]: Alright, this was extra code added after initial PR was made when a feature was requested to list all missing modules rather than one at a time. Is that additional feature no longer a requirement to land?

File: garak/_plugins.py (hunk 31 above)
Comment (leondz) [reply garak/_plugins.py]: We should 

1. merge these threads
2. get clear about reqs for this PR

I don't mind if we advise all missing modules at once or piecemeal. The latter has better UX. Agree this feature should belong in the right PR and code location, if the feature is going to manifest.

Comment (jmartin-tech) [reply garak/_plugins.py]: Understood, based on the targeted changes in scope this PR should remove this block.

Also the revisions already to made to `load_deps` will provide a consistent verbose list of all packages required for the plugin that is attempting to load.  This covers the same scope as what this block does with added context available.

**Review by jmartin-tech — APPROVED** [review]
